### PR TITLE
database: Emit event notifications

### PR DESCRIPTION
**Pull request by swenson**
Including for failures to write credentials and failure to rotate.

### CODE REVIEW

**Review by fairclothjm — APPROVED** [review]
LGTM!

File: builtin/logical/database/path_rotate_credentials.go [hunk 1]
```diff
@@ -75,8 +75,16 @@ func pathRotateRootCredentials(b *databaseBackend) []*framework.Path {
...
 
 func (b *databaseBackend) pathRotateRootCredentialsUpdate() framework.OperationFunc {
-	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (resp *logical.Response, err error) {
 		name := data.Get("name").(string)
+		defer func() {
+			if err == nil {
+				b.dbEvent(ctx, "rotate-root", req.Path, name, true)
+			} else {
+				b.dbEvent(ctx, "rotate-root-fail", req.Path, name, false)
```
Comment (tomhjp) [inline builtin/logical/database/path_rotate_credentials.go:84]: The failure modes are unlikely, but I think it should be possible for modified to be true in this case, if for example the write to storage fails after the password successfully got updated. I think the same is true for a couple of other APIs as well.

File: builtin/logical/database/backend_test.go [hunk 2]
```diff
@@ -585,6 +600,34 @@ func TestBackend_basic(t *testing.T) {
 			t.Fatalf("Creds should not exist")
 		}
 	}
+	assert.Equal(t, 9, len(eventSender.Events))
+	assert.Equal(t, "database/config-write", string(eventSender.Events[0].Type))
```
Comment (tomhjp) [inline builtin/logical/database/backend_test.go:604]: nit: Would be nice to use a for loop to reduce the copy-paste/visual noise here :)

**Review by tomhjp — APPROVED** [review]
LGTM, just a couple of nits

**Comment by swenson** [comment]
Thanks!

**Pull request merged**